### PR TITLE
Finish Variant Class

### LINKED ISSUE #2176: Add `Variant.decode` to Python API
We now have a random access method for variants at a site thanks to `tsk_variant_decode`. I assume we should bring this up to the documented Python API (It is currently undocumented). The documentation of this should be careful to detail its performance characteristics. We should also add/update tutorial content on how to access genotypes.
Also document the `Variant` constructor as part of this.

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #2176

### CODE REVIEW

**Comment by benjeffery** [comment]
Getting some codecov weirdness here, I'm sure the lines are covered that it says aren't.

File: python/tests/test_lowlevel.py [hunk 1]
```diff
@@ -2364,6 +2364,30 @@ def test_alleles(self):
...
+        v = _tskit.Variant(ts, samples=[4, 2], alleles=None)
+        assert np.array_equal(v.samples, np.array([4, 2], dtype=np.int32))
+        v = _tskit.Variant(ts, samples=[], alleles=None)
+        assert np.array_equal(v.samples, np.array([], dtype=np.int32))
+        with pytest.raises(AttributeError):
+            v.samples = [1]
+
+    def test_isolated_as_missing(self):
+        ts = self.get_example_tree_sequence()
+        v = _tskit.Variant(
```
Comment (jeromekelleher) [inline python/tests/test_lowlevel.py:2382]: Odd formatting here?

File: python/tskit/genotypes.py [hunk 2]
```diff
@@ -26,16 +26,23 @@
...
-    mapping sample IDs to the observed alleles.
+    among samples for a given site. A variant consists of (a) the **alleles** that
+    may be observed at the sample nodes for this site; (b) the **genotypes**
+    mapping sample IDs to the observed alleles and (c) of a reference to
+    the :class:`Site` that the Variant has been decoded at.
+
+    After creation a Variant is not yet decoded, and has no genotypes.
+    To decode a Variant, call the :meth:`decode` method. The Variant class will then
+    use a Tree, internal to the Variant, to seek to the position of the site and decode
+    the genotypes at that site. As such it is currently much more efficient to decode
```
Comment (jeromekelleher) [inline python/tskit/genotypes.py:44]: I think it'll always be more efficient to do it this way, indexes will just make it less inefficient to randomly seek. I'd say "It is therefore much more efficient to sites in sequential genomic order, either in a forwards or backwards direction, than to do so randomly".

File: python/tskit/genotypes.py [hunk 3]
```diff
@@ -102,21 +138,40 @@ def alleles(self) -> tuple:
...
+        return self._ll_variant.isolated_as_missing
+
     @property
     def has_missing_data(self) -> bool:
         """
         True if there is missing data for any of the
         samples at the current site.
         """
-        return self._ll_variant.alleles[-1] is None
+        return (
```
Comment (jeromekelleher) [inline python/tskit/genotypes.py:172]: This is creating the ``alleles`` tuple twice I guess? This gets called a lot, so makes sense to keep it tight.

**Review by jeromekelleher — APPROVED** [review]
LGTM, few minor comments.

**Pull request merged**